services/s3/src/main/java/software/amazon/awssdk/services/s3/internal/handlers/SyncChecksumValidationInterceptor.java

```diff
@@ -109,7 +117,17 @@ public void afterUnmarshalling(Context.AfterUnmarshalling context, ExecutionAttr
         }
     }
 
-    private boolean checksumValidationEnabled(ExecutionAttributes executionAttributes) {
+    private boolean checksumValidationEnabled(ExecutionAttributes executionAttributes, Map<String, List<String>> headers) {
+
+        // S3 doesn't support trailing checksums for customer encryption
+        if (headers.containsKey(SERVER_SIDE_CUSTOMER_ENCRYPTION_HEADER)) {
+            return false;
+        }
+
+        // S3 doesn't support trailing checksums for KMS encrypted objects
+        if (headers.getOrDefault(SERVER_SIDE_ENCRYPTION_HEADER, new ArrayList<>()).contains(AWS_KMS.toString())) {
+            return false;
```

Comment:
Should we move the checks to an util method so that it can be used in both sync and async?